### PR TITLE
allow customizable caching permissions

### PR DESCRIPTION
**Pull request by timifasubaa**
This PR allows deployments to restrict access to the cache with their custom logic by extending the new SupersetSecurityManager

@mistercrunch @john-bodley 
cc @fabianmenges I remember George Siros also expressing interest in this functionality.

### CODE REVIEW

**Comment by mistercrunch** [comment]
I think this results in double-checking the permission at the moment, which is an extra database round trip. I'm not sure how efficient it is either for non-all-datasource-access roles, meaning we'd pay a relatively expensive cost twice.

Currently the view is in charge of doing data access checks. Since the auth scheme here is a FAB (web framework) construct, it could be good to keep those check shallow in the stack, meaning in the view layer.

Curious to understand your use case and how the check on cache would differ from the datasource check.

File: superset/security.py [hunk 1]
```diff
@@ -110,6 +110,10 @@ def schema_access(self, datasource, user=None):
             self.can_access('schema_access', datasource.schema_perm, user=user)
         )
 
+    def has_cache_access(self, datasource):
+        # By default datasource access implies cache access.
+        return self.datasource_access(datasource)
```
Comment (john-bodley) [inline superset/security.py]: Could we make this return `True` by default. That's a no-op and the custom security manager could override this.

File: superset/viz.py [hunk 2]
```diff
@@ -309,7 +309,8 @@ def get_df_payload(self, query_obj=None):
         stacktrace = None
         df = None
         cached_dttm = datetime.utcnow().isoformat().split('.')[0]
-        if cache_key and cache and not self.force:
+        can_use_cache = cache_key and cache and not self.force
+        if security_manager.has_cache_access(self.datasource) and can_use_cache:
```
Comment (john-bodley) [inline superset/viz.py]: This logic should all reside in a single if statement, order from cheapest to most expensive, i.e., we should have `security_manager.has_cache_access(self.datasource)` evaluated last. 

As a side note I wonder if we should check the cache timeout here as well, i.e., for Druid datasources one would probably want a timeout of 0 which would imply don't cache.

**Comment by john-bodley** [comment]
@fabianmenges option 2 is what we're thinking. I made a note which describes how we could bypass double-checking the datasource permissions by default.

File: superset/viz.py (hunk 2 above)
Comment (john-bodley) [reply superset/viz.py]: Also I feel deciphering what is going on with the cache is getting somewhat complicated. Maybe a class on connivence methods would help provide more clarity.

**Comment by mistercrunch** [comment]
Why late-checking deep in the code as opposed to in the view layer?

**Comment by timifasubaa** [comment]
I thought about this some more. you're right. With the new approach we are taking now, anyone who can access the datasource can access the cache. I'll go ahead to close this PR.

**Pull request closed without merging**